### PR TITLE
Use only lower floats for Vector3 dot and equality

### LINKED ISSUE #8220: Vector3.Normalize(), .Distance(), and .Equals() do not return correct values in some cases
I've been trying to track down an obscure issue with a project I am working on, and I believe I've found the root cause to be related to `Vector3.Normalize` not returning correct values. I haven't yet been able to get a very minimal repro, but the problem always occurs in the context of my project (which is fairly compex). The scenario is as follows:

* Running an anonymous delegate with one parameter, `ref ParticleState`, where `ParticleState` has a `Vector3 Offset` field (among other stuff).
* Subtract `ParticleState.Offset` from another `Vector3`, retrieved from some other object.
* Call `Vector3.Normalize` on the result of the subtraction.
* Use that normalized direction vector later in the function.

The actual code can be found here: https://github.com/mellinoe/ge-test-game/blob/0939e2496a1cca1229d504643055a4de3c1ce932/code/src/GravityGame/ParticleVacuumEffect.cs#L49

I did some experiments and have a branch [here](https://github.com/mellinoe/ge-test-game/blob/simd-repro/code/src/GravityGame/ParticleVacuumEffect.cs#L49) that prints out info about the problem. The gist of it is that manually computing the normalized difference vector gives a different result than using `Vector3.Normalize(a - b)`. Also, the `.Length()` of values returned from `Vector3.Normalize()` are not always equal to 1, which should be impossible.

```CSharp
// SIMD BUG
Vector3 directionBad = Vector3.Normalize(_vacuumTarget.Position - ps.Offset); // This should be equivalent to "direction" on the next two lines.
Vector3 direction = new Vector3(_vacuumTarget.Position.X - ps.Offset.X, _vacuumTarget.Position.Y - ps.Offset.Y, _vacuumTarget.Position.Z - ps.Offset.Z);
direction = direction / direction.Length();
```
`directionBad != direction`, and `directionBad.Length() != 1`.

Setting `COMPLUS_FEATURESIMD=0` causes the problem to no longer repro. When that is set, the two values above (`directionBad` and `direction`) are identical in all cases.

The issue is also very context-sensitive. When I assign some intermediate results to variables, it no longer repros. For example, if I assign the `_vacuumTarget.Position - ps.Offset` result to a variable, and call `Vector3.Normalize` on that, then the problem doesn't repro. I also use `Vector3.Normalize` in a ton of other places in this project and have not seen issues like this. I suspect it might have something to do with the ref parameter to a struct, or something like that.

I can help investigate and get a repro set up if needed, since I'm not able to find a simpler repro right now. I've been able to repro this on both the 1.0 and 1.1 CoreCLR runtimes, on Windows. I haven't seen if this repros on other OS's, although I could try to check that on Monday.

@CarolEidt , @sivarv

### PR DESCRIPTION
**Pull request by CarolEidt**
For both dot product and comparisons that produce a boolean result, we need to use only the lower 3 floats. The bug was exposed by a case where the result of a call was being used in one of these operations without being stored to a local (which would have caused the upper bits to be cleared).

Fix #8220

### CODE REVIEW

**Comment by CarolEidt** [comment]
@sivarv @dotnet/jit-contrib PTAL

File: src/jit/simd.cpp [hunk 1]
```diff
@@ -754,7 +754,7 @@ SIMDIntrinsicID Compiler::impSIMDLongRelOpEqual(CORINFO_CLASS_HANDLE typeHnd,
     //
     // Equality(v1, v2):
     // tmp = (v1 == v2) i.e. compare for equality as if v1 and v2 are vector<int>
-    // result = BitwiseAnd(t, shuffle(t, (2, 3, 1 0)))
+    // result = BitwiseAnd(t, shuffle(t, (2, 3, 0, 1)))
     // Shuffle is meant to swap the comparison results of low-32-bits and high 32-bits of respective long elements.
```
Comment (CarolEidt) [inline src/jit/simd.cpp]: Note that here, and in the code below, I have simply changed the comment and the name of the shuffle constant to correctly reflect the actual shuffle that was (already) being done. I discovered this as I was changing the dot product code to use shuffle constants to make it easier to follow what is being done.

**Comment by sivarv** [comment]
:shipit:

**Comment by CarolEidt** [comment]
@sivarv I modified the test case because it was failing on x86 without intrinsics, due to result mismatches in the last place. I created a checking method that approximates an ULP (may allow as much as 2 ULP difference), and verified that it still exposes the original bug without my changes. Could you have a look at the test case (especially the `Check` method)? I'd like a second pair of eyes on it.

**Comment by mikedn** [comment]
> I modified the test case because it was failing on x86 without intrinsics

Was that with the legacy x86 JIT? Or in other words, was that caused by a difference between x87 FPU and SSE or between scalar SSE and vector SSE versions?

**Comment by CarolEidt** [comment]
> Was that with the legacy x86 JIT? Or in other words, was that caused by a difference between x87 FPU and SSE or between scalar SSE and vector SSE versions?

I confess that I haven't tracked this down thoroughly (in the interest of moving on to other things), but it was with RyuJIT x86 (without the SIMD intrinsics). At first glance it seemed that it should be doing exactly the same computation, but it *may* be the case (like I say, I haven't done due diligence here), that it is due to the x86 calling convention where we have to return things on the floating point stack, requiring some conversions that might not exist in the code generation for the (seemingly identical) other path.

**Comment by sivarv** [comment]
@CarolEidt - I  am thinking that this precision difference could be an issue for x86 RyuJIT users if indeed there is any.  So it would be good to understand it.  Meanwhile for you to make progress with this PR, I would suggest that we disable this test for x86 with an issue tracking it.

**Comment by mikedn** [comment]
>  that it is due to the x86 calling convention where we have to return things on the floating point stack, requiring some conversions that might not exist in the code generation for the (seemingly identical) other path.

Ah, yes, the FPU is (unfortunately) still needed/used on x86. Anyway, now that I know it happens with RyuJIT x86 I can check myself if you don't have time. Thanks for the info!

**Comment by CarolEidt** [comment]
@sivarv I am disinclined to disable the test, as it is a valuable test, even in the x86 case and I have verified that I am getting results within 1 ULP. I think it is better to have the test as-is than to disable it.

**Comment by CarolEidt** [comment]
@sivarv I propose to add an issue to determine why the results differ for x86 without intrinsics, and if the test can be tweaked to reliably produce identical answers, change the test. But until then, leave the test as-is. Note that this is unlikely to be the last time that we have a test that compares two implementations of something that cannot be guaranteed to produce identical results. It is just not something one should expect when dealing with floating point.

**Comment by sivarv** [comment]
@CarolEidt - I am fine with either approach as long as we have a tracking issue.

**Comment by sivarv** [comment]
Looks good to me.

**Comment by sivarv** [comment]
The only reason I thought it is worth investigation is to understand whether x86 RyuJIT approach of floating point codegen and return via x87 FPU can lead to mismatch in precisions more often than Legacy  x86 (that sticks entirely to x87) or amd64 (that sticks entirely to SSE2).

I might be a good idea to know whether the original test fails due to precision difference either on legacy x86 or amd64 without SIMD intrinsics.

**Comment by CarolEidt** [comment]
Added issue #8506

**Pull request merged**